elastic/apm-pipeline-library
# feat: add junit2otel step supporting traces

## What does this PR do?
It overrides the junit step, supporting passing traces if desired.

The traces are covered by a feature flag i the form of the `JUNIT_2_OTLP` variable. If not empty, the code will evaluate the traces. In any case, at the end of the step it will execute the default junit step, passing all arguments to it.

The traces will be sent using https://github.com/mdelapenya/junit2otlp CLI.

To process the traces, the step:

1. will set the service name, service version and trace name from the environment:
  - If the REPO variable is set, but the OTEL_SERVICE_NAME is not, then REPO will be used as service name. Otherwise OTEL_SERVICE_NAME will be used, and `junit2otel` in the case none of the variables are set.
  - If the REPO variable is set, but the JUNIT_OTEL_TRACE_NAME is not, then REPO will be used as trace name. Otherwise JUNIT_OTEL_TRACE_NAME will be used, and `junit2otel` in the case none of the variables are set.
  - If the JUNIT_OTEL_SERVICE_VERSION is not set, then it will use, in this particular order: pull-request ID, tag name and branch name. Else, it will use 'unknown'.
2. will call a shell script to avoid overloading the jenkins instance
3. the shell script will capture the glob expression representing the `testResults` attribute from the default junit step
4. it will loop over those files, executing the CLI tool as a Docker image.



## Why is it important?
We want to start sending traces for test reports



## How to test the OpenTelemetry traces
I was able to run the shell script having configured Jaeger previously:

1. Prepare Jaeger and OTLP exporter:
```shell
git clone https://github.commdelapenya/junit2otlp.git && cd junit2otlp
make demo-start-jaeger
```

2. Prepare environment to connect to Jaeger from the CLI tool:
```
export OTEL_EXPORTER_OTLP_INSECURE="true"
export OTEL_EXPORTER_OTLP_METRIC_INSECURE="true"
export OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:14317"
export OTEL_EXPORTER_OTLP_SPAN_INSECURE="true"
export OTEL_EXPORTER_OTLP_HEADERS=""
```

3. Run the shell script:
```
$ SERVICE_NAME="apm-pipeline" SERVICE_VERSION="master" TRACE_NAME="junit-tests" TEST_RESULTS_GLOB="/Users/mdelapenya/sourcecode/src/github.com/mdelapenya/junit2otlp/TEST-*" ./resources/scripts/junit2otel.sh 
```

4. Check results on Jaeger: open http://localhost:16686/, and traces for the `apm-pipeline` service should be displayed, using `junit-tests` as trace name and `master` as version.

<img width="1792" alt="Screenshot 2022-01-26 at 18 17 02" src="https://user-images.githubusercontent.com/951580/151213314-f550fdd3-457f-4257-886e-2b4bbae7d255.png">

## Related issues
- Supersedes https://github.com/elastic/apm-pipeline-library/pull/1368
